cmd/project.go

```diff
@@ -292,49 +312,52 @@ func listProjectRoles(ctx *cli.Context) error {
 	return listRoles(ctx, "project")
 }
 
-func listProjectMembers(ctx *cli.Context) error {
-	c, err := GetClient(ctx)
-	if err != nil {
-		return err
-	}
+type prtbLister interface {
+	List(opts *types.ListOpts) (*managementClient.ProjectRoleTemplateBindingCollection, error)
+}
 
-	projectID := c.UserConfig.Project
+func listProjectMembers(ctx *cli.Context, out io.Writer, config userConfig, prtbs prtbLister, principals principalGetter) error {
```

Comment:
I love the change of the signature to make tests easier (and possible). I would try to avoid passing around the `cli.Context` if not really needed, and pass the needed arguments instead. I.e. something like a more generic `ListProjectMembersConfig` with the possible configuration. Not sure if we need a userConfig, or we can just put everything in one.

```go
type ListProjectMembersConfig struct {
	ProjectID	string
	Format		bool
	Quiet		bool
}

listProjectMembers(config ListProjectMembersConfig, out io.Writer, prtbs prtbLister, principals principalGetter) {
	// do stuff
}
```

Maybe we can even put the whole `TableWriterConfig` in the Configc, and it will be cleaner:

```go
type ListProjectMembersConfig struct {
	ProjectID		string
	TableWriterConfig	*TableWriterConfig
}

listProjectMembers(config ListProjectMembersConfig, prtbs prtbLister, principals principalGetter) {
	// do stuff
}
```

We could also have a constructor, helper and validation to create the config from the cli.Context, if needed.

What do you think?

Reply:
This is a good idea, but I would suggest we look holistically at all commands and see what the best way to refactor this. At the moment and for the purposes of this PR `cli.Context` is as good of a container as any struct to pass flags around. There is certain amount of common flags that are being handled like default list and format options that need to considered.

Reply:
That's good to me. Let's try to not diverge too much. :)